File: src/weaponsprocurement/gui/StockReviewFilterState.java
package weaponsprocurement.gui;

import java.util.EnumMap;
import java.util.EnumSet;
import java.util.Map;
import java.util.Set;

final class StockReviewFilterState {
    private final Set<StockReviewFilter> activeFilters = EnumSet.noneOf(StockReviewFilter.class);
    private final Map<StockReviewFilterGroup, Boolean> expandedFilterGroups =
            new EnumMap<StockReviewFilterGroup, Boolean>(StockReviewFilterGroup.class);

    StockReviewFilterState() {
        initializeFilterGroups();
    }

    StockReviewFilterState(StockReviewFilterState source) {
        activeFilters.addAll(source.activeFilters);
        expandedFilterGroups.putAll(source.expandedFilterGroups);
        ensureFilterGroupsInitialized();
    }

    boolean isFilterActive(StockReviewFilter filter) {
        return activeFilters.contains(filter);
    }

    void toggleFilter(StockReviewFilter filter) {
        if (filter == null) {
            return;
        }
        if (activeFilters.contains(filter)) {
            activeFilters.remove(filter);
        } else {
            activeFilters.add(filter);
        }
    }

    Set<StockReviewFilter> getActiveFilters() {
        return EnumSet.copyOf(activeFilters);
    }

    int getActiveFilterCount() {
        return activeFilters.size();
    }

    void clearFilters() {
        activeFilters.clear();
    }

    boolean isExpanded(StockReviewFilterGroup group) {
        Boolean value = expandedFilterGroups.get(group);
        return value != null && value.booleanValue();
    }

    void toggle(StockReviewFilterGroup group) {
        if (group != null) {
            expandedFilterGroups.put(group, Boolean.valueOf(!isExpanded(group)));
        }
    }

    private void initializeFilterGroups() {
        for (StockReviewFilterGroup group : StockReviewFilterGroup.values()) {
            expandedFilterGroups.put(group, Boolean.FALSE);
        }
    }

    private void ensureFilterGroupsInitialized() {
        for (StockReviewFilterGroup group : StockReviewFilterGroup.values()) {
            if (!expandedFilterGroups.containsKey(group)) {
                expandedFilterGroups.put(group, Boolean.FALSE);
            }
        }
    }
}

File: src/main/kotlin/weaponsprocurement/gui/StockReviewFilterState.kt
package weaponsprocurement.gui

import java.util.EnumMap
import java.util.EnumSet

class StockReviewFilterState {
    private val activeFilters: EnumSet<StockReviewFilter> = EnumSet.noneOf(StockReviewFilter::class.java)
    private val expandedFilterGroups: EnumMap<StockReviewFilterGroup, Boolean> =
        EnumMap(StockReviewFilterGroup::class.java)

    constructor() {
        initializeFilterGroups()
    }

    constructor(source: StockReviewFilterState) {
        activeFilters.addAll(source.activeFilters)
        expandedFilterGroups.putAll(source.expandedFilterGroups)
        ensureFilterGroupsInitialized()
    }

    fun isFilterActive(filter: StockReviewFilter?): Boolean = activeFilters.contains(filter)

    fun toggleFilter(filter: StockReviewFilter?) {
        if (filter == null) {
            return
        }
        if (activeFilters.contains(filter)) {
            activeFilters.remove(filter)
        } else {
            activeFilters.add(filter)
        }
    }

    fun getActiveFilters(): Set<StockReviewFilter> = EnumSet.copyOf(activeFilters)

    fun getActiveFilterCount(): Int = activeFilters.size

    fun clearFilters() {
        activeFilters.clear()
    }

    fun isExpanded(group: StockReviewFilterGroup?): Boolean = expandedFilterGroups[group] == true

    fun toggle(group: StockReviewFilterGroup?) {
        if (group != null) {
            expandedFilterGroups[group] = !isExpanded(group)
        }
    }

    private fun initializeFilterGroups() {
        for (group in StockReviewFilterGroup.values()) {
            expandedFilterGroups[group] = false
        }
    }

    private fun ensureFilterGroupsInitialized() {
        for (group in StockReviewFilterGroup.values()) {
            if (!expandedFilterGroups.containsKey(group)) {
                expandedFilterGroups[group] = false
            }
        }
    }
}
